Add ability to authenticate multiple domains

PackeTsar opened the pull request:
# Issue
The current plugin does not allow multiple domains to be authenticated.

This is because when multiple domains are specified in the invocation of certbot, certbot will make multiple independent calls to `_perform()`. Each time `_perform()` is called, it calls `add_txt_record()` which checks for an existing record and updates/changes that record if it does indeed exist. 

These updates to the existing TXT record result in a single **_acme-challenge.example.com** TXT record with a changed value. In order for Let's Encrypt to authenticate multiple domains, it requires multiple TXT records to exist all with the same (_acme-challenge.example.com) name.



# New Code
The new code does not overwrite existing TXT records with the _acme-challenge.example.com name. Instead the new code pulls a list of the existing records with that name, and inserts the new record in addition to them. It does this by combining all the records into a list of records and issuing a PATCH against the API with all the records passed in the payload.

The new code also performs an iterative delete of all the records with the _acme-challenge.example.com name once certbot calls the `_cleanup()` method.

helgeerbe commented:
Hi @PackeTsar, Thanks for providing this pull request. I wasn't aware IONOS is used outside of Europe. I don't use multiple domains, but I will check it and merge it to master.

PackeTsar commented:
Yea no problem at all. IONOS used to be "1&1" in the US. And API access requires a special request (which is a pain to do). I didn't run the test suite against this new code, but when I ran it before it seemed to be broken, so I didn't bother after the modifications.

helgeerbe commented:
Hi @PackeTsar,

I pulled your code and tried to understand what exactly it does. It's been a while I wrote this plug in.

In my understanding you are going to retrieve *one* certificate for `server.example1.com` which has an alternative domain like `server.example2.com`?

For that you create the txt_record and patch this in appending the second domain?

PackeTsar commented:
Yes the new code allows the creation of a certificate with multiple domains in it. In my case, I needed `*.example.com` as well as `example.com` to be in the cert. In order to authenticate for this type of cert, the certbot makes multiple calls to the plugin with different validation values.

The old code would end up overwriting the same TXT record with the multiple values, resulting in a failed authentication attempt.

The new code adds TXT records as they are passed in by certbot, and then removes them all at the end to clean up. This results in successful authentication for the multiple-domain certificate.

PackeTsar commented:
Yea I'm not sure I follow what you are seeing in the logs, can you post your sanitized log here?

Regarding your "update txt record" function reusing the record, you are correct there. The reuse of that record is what would break authentication. Below is a step by step of the process for the two versions of code.

# Old Code
1. Certbot is run and is passed multiple domains to put in the cert
2. Certbot queries Let's Encrypt (LE) and obtains two authentication tokens to use, one for each domain
3. Certbot calls the `_perform()` function with the *first* auth token to have the DNS record created
  a. `perform()` calls `add_txt_record()` to add the record
  b. `add_txt_record()` uses `get_existing_txt()` to check for the pre-existence of the record
  c. The record *does not* exist, so `add_txt_record()` calls `_insert_txt_record()` to add the record
  d. The plugin returns control back to Certbot
4. Certbot calls the `_perform()` function with the *second* auth token to have the DNS record created
  a. `perform()` calls `add_txt_record()` to add the record
  b. `add_txt_record()` uses `get_existing_txt()` to check for the pre-existence of the record
  c. The record *does* exist, so `add_txt_record()` calls `_update_txt_record()` to update its values
  d. `_update_txt_record()` overwrites the first record (with the first token's data) with the second token's data
  e. The plugin returns control back to Certbot
5. Certbot cues LE to check DNS and finish the authentication
6. LE queries DNS but only finds the value of the second token, so the first token (for the first domain) doesn't pass
7. Certbot aborts and fails the ACME process


# New Code
1. Certbot is run and is passed multiple domains to put in the cert
2. Certbot queries Let's Encrypt (LE) and obtains two authentication tokens to use, one for each domain
3. Certbot calls the `_perform()` function with the *first* auth token to have the DNS record created
  a. `perform()` calls `add_txt_record()` to add the record
  b. `add_txt_record()` uses `get_existing_txt()` to check for the pre-existence of the record
  c. The record *does not* exist, so `add_txt_record()` calls `_insert_txt_record()` to add the record
  d. The plugin returns control back to Certbot
4. Certbot calls the `_perform()` function with the *second* auth token to have the DNS record created
  a. `perform()` calls `add_txt_record()` to add the record
  b. `add_txt_record()` uses `get_existing_txt()` to check for the pre-existence of the record
  c. The record *does* exist, so `add_txt_record()` uses `get_existing_records()` and `clean_entries()` to gather and clean the existing records
  d. It then calls `add_additional_record()` to append a new record with the new token data alongside the existing ones
  e. The plugin returns control back to Certbot
5. Certbot cues LE to check DNS and finish the authentication
6. LE queries DNS and finds all the TXT records it expects to be there, so it authorizes the cert creation
7. Certbot finishes and generates the certificate

helgeerbe commented:
Which version of certbot do you use. I did a fresh installation. certbot version: 1.20.0

PackeTsar commented:
I'm using 1.19.0. Here are my logs

```
...removed...

2021-10-09 18:29:55,414:DEBUG:acme.client:Storing nonce: <private>
2021-10-09 18:29:55,414:INFO:certbot._internal.auth_handler:Performing the following challenges:
2021-10-09 18:29:55,415:INFO:certbot._internal.auth_handler:dns-01 challenge for example.com
2021-10-09 18:29:55,415:INFO:certbot._internal.auth_handler:dns-01 challenge for example.com
2021-10-09 18:29:55,415:WARNING:certbot.plugins.dns_common:Unsafe permissions on credentials configuration file: /opt/ionos_api_key.ini
2021-10-09 18:29:55,417:DEBUG:certbot_dns_ionos.dns_ionos:_perform called with: domain: example.com, validation_name: _acme-challenge.example.com, validation: <private>
2021-10-09 18:29:55,417:DEBUG:certbot_dns_ionos.dns_ionos:creating ionosclient
2021-10-09 18:29:55,417:DEBUG:certbot_dns_ionos.dns_ionos:get zones
2021-10-09 18:29:55,418:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:30:28,848:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "GET /dns/v1/zones HTTP/1.1" 200 3668
2021-10-09 18:30:28,849:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones
2021-10-09 18:30:28,850:DEBUG:certbot_dns_ionos.dns_ionos:zones found [{'name': 'example.com', 'id': '<private-id>', 'type': 'NATIVE'}, ...truncated]
2021-10-09 18:30:28,850:DEBUG:certbot_dns_ionos.dns_ionos:domain found: example.com with id: <private-id>
2021-10-09 18:30:28,851:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:31:01,611:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "GET /dns/v1/zones/<private-id> HTTP/1.1" 200 54311
2021-10-09 18:31:01,801:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones/<private-id>
2021-10-09 18:31:01,802:INFO:certbot_dns_ionos.dns_ionos:insert new txt record
2021-10-09 18:31:01,802:DEBUG:certbot_dns_ionos.dns_ionos:insert with data: {'disabled': False, 'type': 'TXT', 'name': '_acme-challenge.example.com', 'content': '<private>', 'ttl': 60, 'prio': 0}
2021-10-09 18:31:01,803:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:31:34,543:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "PATCH /dns/v1/zones/<private-id> HTTP/1.1" 200 0
2021-10-09 18:31:34,545:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones/<private-id>
2021-10-09 18:31:34,547:DEBUG:certbot_dns_ionos.dns_ionos:_perform called with: domain: example.com, validation_name: _acme-challenge.example.com, validation: <private-id>
2021-10-09 18:31:34,548:DEBUG:certbot_dns_ionos.dns_ionos:creating ionosclient
2021-10-09 18:31:34,548:DEBUG:certbot_dns_ionos.dns_ionos:get zones
2021-10-09 18:31:34,548:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:32:07,546:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "GET /dns/v1/zones HTTP/1.1" 200 3668
2021-10-09 18:32:07,547:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones
2021-10-09 18:32:07,547:DEBUG:certbot_dns_ionos.dns_ionos:zones found [{'name': 'example.com', 'id': '<private>', 'type': 'NATIVE'}, ...truncated]
2021-10-09 18:32:07,547:DEBUG:certbot_dns_ionos.dns_ionos:domain found: example.com with id: <private-id>
2021-10-09 18:32:07,548:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:32:40,325:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "GET /dns/v1/zones/<private-id> HTTP/1.1" 200 54578
2021-10-09 18:32:40,517:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones/<private-id>
2021-10-09 18:32:40,517:INFO:certbot_dns_ionos.dns_ionos:adding additional record
2021-10-09 18:32:40,518:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:33:13,047:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "GET /dns/v1/zones/<private-id> HTTP/1.1" 200 54578
2021-10-09 18:33:13,240:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones/<private-id>
2021-10-09 18:33:13,240:DEBUG:certbot_dns_ionos.dns_ionos:insert with data: [{'name': '_acme-challenge.example.com', 'type': 'TXT', 'content': '<private>', 'ttl': 60, 'disabled': False}, {'disabled': False, 'type
': 'TXT', 'name': '_acme-challenge.example.com', 'content': '<private-id>', 'ttl': 60, 'prio': 0}]
2021-10-09 18:33:13,241:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:33:45,859:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "PATCH /dns/v1/zones/<private-id> HTTP/1.1" 200 0
2021-10-09 18:33:45,860:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones/<private-id>
2021-10-09 18:33:45,861:DEBUG:certbot._internal.display.obj:Notifying user: Waiting 120 seconds for DNS changes to propagate
2021-10-09 18:35:45,893:DEBUG:acme.client:JWS payload:
b'{}'
2021-10-09 18:35:45,895:DEBUG:acme.client:Sending POST request to https://acme-staging-v02.api.letsencrypt.org/acme/chall-v3/xxxxxxxx/xxxxxxx:

...removed...

2021-10-09 18:36:18,528:DEBUG:acme.client:Storing nonce: <private>
2021-10-09 18:36:18,528:DEBUG:certbot._internal.error_handler:Calling registered functions
2021-10-09 18:36:18,528:INFO:certbot._internal.auth_handler:Cleaning up challenges
2021-10-09 18:36:18,529:DEBUG:certbot_dns_ionos.dns_ionos:creating ionosclient
2021-10-09 18:36:18,529:DEBUG:certbot_dns_ionos.dns_ionos:get zones
2021-10-09 18:36:18,530:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:36:51,083:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "GET /dns/v1/zones HTTP/1.1" 200 3668
2021-10-09 18:36:51,085:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones
2021-10-09 18:36:51,086:DEBUG:certbot_dns_ionos.dns_ionos:zones found [{'name': 'example.com', 'id': '<private>', 'type': 'NATIVE'}, ...truncated]
2021-10-09 18:36:51,086:DEBUG:certbot_dns_ionos.dns_ionos:domain found: example.com with id: <private-id>
2021-10-09 18:36:51,087:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:37:23,639:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "GET /dns/v1/zones/<private-id> HTTP/1.1" 200 54845
2021-10-09 18:37:23,836:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones/<private-id>
2021-10-09 18:37:23,837:DEBUG:certbot_dns_ionos.dns_ionos:delete id: <private-id>
2021-10-09 18:37:23,838:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:37:56,401:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "DELETE /dns/v1/zones/<private-id>/records/<private-id> HTTP/1.1" 200 0
2021-10-09 18:37:56,402:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones/<private-id>/records/<private-id>
2021-10-09 18:37:56,402:DEBUG:certbot_dns_ionos.dns_ionos:delete id: <private-id>
2021-10-09 18:37:56,403:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:38:28,978:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "DELETE /dns/v1/zones/<private-id>/records/<private-id> HTTP/1.1" 200 0
2021-10-09 18:38:28,980:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones/<private-id>/records/<private-id>
2021-10-09 18:38:28,981:DEBUG:certbot_dns_ionos.dns_ionos:creating ionosclient
2021-10-09 18:38:28,981:DEBUG:certbot_dns_ionos.dns_ionos:get zones
2021-10-09 18:38:28,982:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:39:01,410:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "GET /dns/v1/zones HTTP/1.1" 200 3668
2021-10-09 18:39:01,411:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones
2021-10-09 18:39:01,411:DEBUG:certbot_dns_ionos.dns_ionos:zones found [{'name': 'example.com', 'id': '<private>', 'type': 'NATIVE'}, ...truncated]
2021-10-09 18:39:01,411:DEBUG:certbot_dns_ionos.dns_ionos:domain found: example.com with id: <private-id>
2021-10-09 18:39:01,412:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.hosting.ionos.com:443
2021-10-09 18:39:34,051:DEBUG:urllib3.connectionpool:https://api.hosting.ionos.com:443 "GET /dns/v1/zones/<private-id> HTTP/1.1" 200 54311
2021-10-09 18:39:34,244:DEBUG:certbot_dns_ionos.dns_ionos:API REquest to URL: https://api.hosting.ionos.com/dns/v1/zones/<private-id>

...removed...

```

helgeerbe commented:
Which parameters are you using to call certbot?

I do:

`'-a', 'dns-ionos', '-d', '*.erbehome.de', '-d', 'erbehome.de', '--dns-ionos-credentials', 'my_debug/secrets/credentials.ini', '--config-dir', 'my_debug/config', '--work-dir', 'my_debug/work', '--logs-dir', 'my_debug/logs'`
